torch/csrc/cuda/nccl.cpp

```diff
@@ -1158,6 +1162,39 @@ void scatter(
   NCCL_CHECK_TIMEOUT(ncclCommCount(comm, &numranks), _comm);
   NCCL_CHECK_TIMEOUT(ncclCommUserRank(comm, &cur_rank), _comm);
 #endif
+#if ((NCCL_MAJOR > 2) || ((NCCL_MAJOR == 2) && (NCCL_MINOR >= 28)))
+  void* send_ptr = nullptr;
+  auto type = to_nccl_data_type(outputs);
+  int64_t count = outputs.numel();
+  at::Tensor flat; // keep alive until after NCCL call
+  if (cur_rank == root) {
+    TORCH_CHECK(
+        (int)inputs.size() == numranks,
+        "root must provide inputs.size()==numranks");
+    // Allocate one flat buffer [world_size * count]
+    flat = at::empty({numranks * count}, outputs.options());
+
+    // Pack each inputs[i] into its slot
+    for (int i = 0; i < numranks; ++i) {
+      const at::Tensor& src = inputs[i].contiguous().view({count});
+      TORCH_CHECK(
+          src.scalar_type() == outputs.scalar_type(),
+          "dtype mismatch at i=",
+          i);
+      TORCH_CHECK(
+          src.is_cuda() && src.get_device() == outputs.get_device(),
+          "device mismatch at i=",
+          i);
+      flat.narrow(0, i * count, count).copy_(src, /*non_blocking=*/true);
```

Comment:
Is there a layout trick we can to pack the src tensor into copy_ with the right striding here without a possible intermediate copy?